### PR TITLE
feat: action callback results - phase 1

### PR DESCRIPTION
**Pull request by mattmazzola**
#### Is this adding or improving a _feature_ or fixing a _bug_?
Feature: Action Stubs: Phase 1

#### What's the new behavior?
Allows user to choose a stub for a callback action when selecting it during training. This ensures consistent replay and reliable way to simulate outcomes of unstable callbacks.

[Video](https://microsoft.sharepoint.com/:v:/t/convsysresearch/Ef6b5aluOfNLg3jGYjWRc-QBYOEG8OLAw8n8QkMI-f7Okw?e=C7UT1x)

#### How does this change work?
Adds new `stubs` property to callback definitions. Similar to root, they can give a name to identify the outcome and provide a `logic` or `render` methods to make the necessary changes to entities.

Adds dropdown to select stub in action scorer
image

Adds stubs on action modal
image

### CODE REVIEW

File: packages/samples/src/demos/tutorialAPICalls.ts [hunk 1]
```diff
@@ -199,6 +199,50 @@ cl.AddCallback({
     }
 })
 
+cl.AddCallback({
+    name: 'MyCallbackWithStubs',
```
Comment (mattmazzola) [inline packages/samples/src/demos/tutorialAPICalls.ts]: Might have to remove before merging until we decide plan for tutorial. Would need to add entities to support it.

File: packages/ui/src/components/modals/TeachSessionModal.tsx [hunk 2]
```diff
@@ -486,7 +486,6 @@ class TeachModal extends React.Component<Props, ComponentState> {
 
     @autobind
     renderSelectedActivity(activity: BB.Activity): (JSX.Element | null) {
```
Comment (mattmazzola) [inline packages/ui/src/components/modals/TeachSessionModal.tsx]: There isn't change here to mimic EditDialogModal because there isn't a train dialog to get the stubName chosen. It doesn't seem too be detrimental though because there are different use cases which doesn't rely on visualizing it in the activity.

When teaching you already know the stub chosen because you just chose it using the dropdown. When editing you are reviewing what happened which is why it was more important.

File: packages/models/src/Callback.ts [hunk 3]
```diff
@@ -13,4 +13,13 @@ export interface Callback {
...
   renderArguments: string[]
   isRenderFunctionProvided: boolean
+  results: CallbackResult[]
 }
+
+export type EntityValue = string | number | boolean | object
+
+export interface CallbackResult {
+  name: string
+  entityValues: Record<string, EntityValue | EntityValue[] | null | undefined>
```
Comment (tpmsr) [inline packages/models/src/Callback.ts]: Do you need `string` in the type list since `EntityValue` can take a `string` ?

File: packages/models/src/Callback.ts [hunk 4]
```diff
@@ -13,4 +13,13 @@ export interface Callback {
   isLogicFunctionProvided: boolean
   renderArguments: string[]
   isRenderFunctionProvided: boolean
+  results: CallbackResult[]
```
Comment (tpmsr) [inline packages/models/src/Callback.ts]: I have a strong objection to the identifier `results`.  I don't think this at all communicates the idea that the results are static / mock values for use in generating deterministic dialog replay.

A few alternative ideas :

- mockResults
- stubResults
- resultsForReplay
- staticReplayResults

File: packages/samples/src/test/testApiCallbacks.ts [hunk 5]
```diff
@@ -271,6 +271,113 @@ cl.AddCallback({
...
+        }
+
+        return randomNumber
+    },
+    render: async n => {
+        return `My random number is ${n}`
+    },
+    results: [
+        {
+            name: 'Number is 3',
```
Comment (tpmsr) [inline packages/samples/src/test/testApiCallbacks.ts]: What do you think about just stringifying the value of this object rather than providing an explicit `name` ?  That would make the name never go stale, but makes the assumption that each stub will only care about changing a small number of properties.

File: packages/samples/src/demos/tutorialAPICalls.ts [hunk 6]
```diff
@@ -200,6 +200,7 @@ cl.AddCallback({
 })
 
 
+
```
Comment (tpmsr) [inline packages/samples/src/demos/tutorialAPICalls.ts]: nit: revert this change

File: packages/ui/src/components/modals/ActionCallbackResultDropdown.tsx [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            key: callbackResult.name,
+            text: callbackResult.name,
+            data: callbackResult,
+        }))
+
+        callbackResultOptions.push(...callbackResultOptionsFromBot)
+    }
+
+    const [selectedCallbackResultOptionKey, setSelectedCallbackResultOptionKey] = React.useState(props.selectedCallbackResult?.name ?? noneOptionKey)
+    // React.useEffect(() => {
```
Comment (tpmsr) [inline packages/ui/src/components/modals/ActionCallbackResultDropdown.tsx]: nit : if this is going to be used in the near future, please add a comment.  Otherwise, please delete.

File: packages/models/src/Callback.ts (hunk 3 above)
Comment (mattmazzola) [reply packages/models/src/Callback.ts]: I'm not sure what you mean, `string` is included in `EntityValue`

File: packages/samples/src/test/testApiCallbacks.ts (hunk 5 above)
Comment (mattmazzola) [reply packages/samples/src/test/testApiCallbacks.ts]: The name provides valuable semantic meaning.  Developer might tell the other user, use the callback result that called "Item is in Stock", regardless of what the entities are named or values are returned.  Similar to your comment this morning about a user looking at an weird ID, it's weird to look at stringified json object

File: packages/models/src/Callback.ts (hunk 4 above)
Comment (mattmazzola) [reply packages/models/src/Callback.ts]: It's not only for replay so bottom two are misleading. User might not even know what replay is and doesn't need to be concerned about that. They might think these are only for replay and disregard them.  It's more important for creating dialogs as we want users to define these up front and use them during training.

In my opinions they are closest to a list of mock logic result objects. Logic result is already used internally the SDK and has concept of entities and return value. But this name is bit weird for users to see.

I think mocks are static objects and stubs are functions.  callbacks provide `logic` and `render` functions and when it was called `stubs` it was a list of stub callbacks. When stub was chosen we call those functions instead of the actual functions.  This is very close to how stubs work everywhere and expected to feel familiar.

Because we're not doing functions anymore so feels wrong to keep calling them stubs. There more like mocks, because they're static objects but it's still feels weird to call them mocks because they're not mocking any real object the user can observe or construct in real callbacks. Thus there is no object to mimic or mock. Real callbacks use memory manager to set and effect things they don't return objects.

Also, it used to represent the whole callback. Now that it's only simulating the logic function.

Will let you and Lars decide.  Already spent too much time on renaming from stub to callback result. Whatever the next name is please pick one and make it final.

File: packages/models/src/Callback.ts (hunk 3 above)
Comment (tpmsr) [reply packages/models/src/Callback.ts]: Yep, see that.  Still had a foggy head yesterday.

File: packages/models/src/Callback.ts (hunk 4 above)
Comment (tpmsr) [reply packages/models/src/Callback.ts]: Just caught up with Lars.  Let's go with `mockResults`.  Thanks!

File: packages/ui/src/components/modals/ActionCallbackResultDropdown.tsx (hunk 7 above)
Comment (mattmazzola) [reply packages/ui/src/components/modals/ActionCallbackResultDropdown.tsx]: Removed

File: packages/models/src/Callback.ts (hunk 4 above)
Comment (mattmazzola) [reply packages/models/src/Callback.ts]: Ok, renamed.

File: packages/sdk/src/CLRunner.ts [hunk 8]
```diff
@@ -1522,9 +1582,39 @@ export class CLRunner {
...
                     // and this causes issues when calculating changedFilledEntities.
                     const entityMapBeforeCall = new CLM.FilledEntityMap(await state.EntityState.FilledEntityMap())
-                    // Store logic callback value
-                    const logicObject = await callback.logic(memoryManager, ...renderedLogicArgumentValues)
-                    logicResult.logicValue = JSON.stringify(logicObject)
+
+                    // If callback result is set use it
+                    // Otherwise use logic function
+                    let logicReturnValue: unknown
+                    if (actionInput.resultName) {
```
Comment (LarsLiden) [inline packages/sdk/src/CLRunner.ts]: As I'm reading this, I thinking "resultName" could actually be a big confusing.   Maybe "mockResultName"?   Matches "mockResults" in Callback.ts

File: packages/sdk/src/CLRunner.ts [hunk 9]
```diff
@@ -1878,7 +1965,7 @@ export class CLRunner {
             }
 
             // Use scorer step to populate pre-built data (when)
-            if (round.scorerSteps && round.scorerSteps.length > 0) {
+            if (round.scorerSteps.length > 0) {
```
Comment (LarsLiden) [inline packages/sdk/src/CLRunner.ts]: Is this safe?  Imagine the null check was added here for a reason.

File: packages/sdk/src/CLRunner.ts [hunk 10]
```diff
@@ -1947,7 +2035,7 @@ export class CLRunner {
                     },
                     labelAction: undefined,
                     logicResult: undefined,
-                    scoredAction: undefined
+                    scoredAction: undefined,
```
Comment (LarsLiden) [inline packages/sdk/src/CLRunner.ts]: Why the comma?

File: packages/sdk/src/CLRunner.ts [hunk 11]
```diff
@@ -1835,26 +1924,24 @@ export class CLRunner {
      * Replay a TrainDialog, calling EntityDetection callback and API Logic,
      * recalculating FilledEntities along the way
      */
-    public async ReplayTrainDialogLogic(trainDialog: CLM.TrainDialog, state: CLState, cleanse: boolean): Promise<CLM.TrainDialog> {
+    public async ReplayTrainDialogLogic(trainDialog: CLM.TrainDialog, state: CLState, cleans: boolean): Promise<CLM.TrainDialog> {
```
Comment (LarsLiden) [inline packages/sdk/src/CLRunner.ts]: Curious why the change from "cleanse" to "cleans"?

Comment (mattmazzola) [reply packages/sdk/src/CLRunner.ts]: Hmm looked like another typo. Didn't realize `cleanse` was spelled that way. Anyways, both the words are meaningless so it doesn't matter, will change it back.

File: packages/sdk/src/CLRunner.ts (hunk 9 above)
Comment (mattmazzola) [reply packages/sdk/src/CLRunner.ts]: Yea I used to think things were added for a reason. Anyways similar to Thomas changes, we'll find out who was breaking the types.

File: packages/sdk/src/CLRunner.ts (hunk 10 above)
Comment (mattmazzola) [reply packages/sdk/src/CLRunner.ts]: Because the language supports it now and it's better.

Comment (LarsLiden) [reply packages/sdk/src/CLRunner.ts]: What's the advantage?  Easier add subsequent items when editing?

File: packages/sdk/src/CLRunner.ts (hunk 9 above)
Comment (LarsLiden) [reply packages/sdk/src/CLRunner.ts]: Removing this adds potential instability to the system (hard failure).   I understanding wanting to find the offending code that passes a bad object, but a console error would be better.

File: packages/sdk/src/CLRunner.ts (hunk 10 above)
Comment (mattmazzola) [reply packages/sdk/src/CLRunner.ts]: Yea, developer productivity. I think mostly code reviews as you get to see the addition of something instead of the misleading edit of line above, but also makes it easier to manipulate text as copying line or moving within list always works.

File: packages/sdk/src/CLRunner.ts (hunk 8 above)
Comment (mattmazzola) [reply packages/sdk/src/CLRunner.ts]: changed to `mockResultName`

File: packages/sdk/src/CLRunner.ts (hunk 9 above)
Comment (mattmazzola) [reply packages/sdk/src/CLRunner.ts]: ok

File: packages/sdk/src/CLRunner.ts (hunk 10 above)
Comment (tpmsr) [reply packages/sdk/src/CLRunner.ts]: +1, love it.  This has been supported in C++ forever.

File: packages/models/src/Callback.ts [hunk 12]
```diff
@@ -13,4 +13,13 @@ export interface Callback {
   isLogicFunctionProvided: boolean
   renderArguments: string[]
   isRenderFunctionProvided: boolean
+  mockResults: CallbackResult[]
```
Comment (tpmsr) [inline packages/models/src/Callback.ts]: Thanks!  Could you please also add a comment describing this field to the interface comments, since we have them ?

File: packages/sdk/src/CLRunner.ts [hunk 13]
```diff
@@ -41,6 +41,34 @@ export enum SessionStartFlags {
 export interface InternalCallback<T> extends CLM.Callback, ICallback<T> {
 }
 
+export const convertInternalCallbackToCallback = <T>(c: InternalCallback<T>): CLM.Callback => {
+    const { logic, render, mockResults, ...callback } = c
+
+    const resultsWithoutUndefinedValues = mockResults.map(r => {
```
Comment (tpmsr) [inline packages/sdk/src/CLRunner.ts]: nit : `resultsWithDefinedValues` would be easier to understand than the double-negative

File: packages/sdk/src/CLRunner.ts [hunk 14]
```diff
@@ -791,6 +821,29 @@ export class CLRunner {
...
+        const callbackInputResults = callbackInput.mockResults
+        if (callbackInputResults) {
+            // Validate result names
+            // Must not equal root/parent condition with actual functions
+            // Must be unique within set of other results
+            const existingResultNames: string[] = []
+            for (const callbackResult of callbackInputResults) {
+                const resultHasNoName = (typeof callbackResult.name !== "string" || callbackResult.name.trim().length === 0)
+                if (resultHasNoName) {
+                    throw new Error(`You attempted to add result on callback ${callbackInput.name} but did not provide a valid name. Name must be non-empty string.`)
```
Comment (tpmsr) [inline packages/sdk/src/CLRunner.ts]: s/result/mockResults/

File: packages/sdk/src/CLRunner.ts [hunk 15]
```diff
@@ -791,6 +821,29 @@ export class CLRunner {
...
+                }
+
+                const resultNameIsNotUnique = callbackResult.name === callbackInput.name
+                    || existingResultNames.includes(callbackResult.name)
+
+                if (resultNameIsNotUnique) {
+                    throw new Error(`You attempted to add a result with the same name, ${callbackInput.name}, as the callback or one of the other results. The results names must be unique.`)
```
Comment (tpmsr) [inline packages/sdk/src/CLRunner.ts]: ditto, mockResults here

File: packages/ui/src/components/modals/ActionCallbackResultDropdown.tsx [hunk 16]
```diff
@@ -0,0 +1,100 @@
+import * as React from 'react'
```
Comment (tpmsr) [inline packages/ui/src/components/modals/ActionCallbackResultDropdown.tsx]: Could you please add a comment to this file describing what it's used for, and from which page/context ?

I think these types of comments would make the UI code a lot easier to navigate for people not as familar with the UI.

Ditto for other new tsx files in the PR.

File: packages/models/src/Callback.ts (hunk 12 above)
Comment (mattmazzola) [reply packages/models/src/Callback.ts]: Yea, but added comment.

Little hesitant about this.  When something is unclear it's usually a sign of bad design. Sure you can add comments to clarify in that one spot, but it would be better to fix the design.

This is why the property should have been named `mockLogicResults` so it was clear exactly what they are.

File: packages/sdk/src/CLRunner.ts (hunk 13 above)
Comment (mattmazzola) [reply packages/sdk/src/CLRunner.ts]: I agree on double negatives being confusing but in this cause it more precisely describes what the code does.  I would not consider `null` a defined value but yet it is included.

Changed to `resultsWithUndefinedAsNull` to hopefully remove negative and be more clear.

File: packages/sdk/src/CLRunner.ts (hunk 14 above)
Comment (mattmazzola) [reply packages/sdk/src/CLRunner.ts]: Add "mock" to error messages

File: packages/ui/src/components/modals/ActionCallbackResultDropdown.tsx (hunk 16 above)
Comment (mattmazzola) [reply packages/ui/src/components/modals/ActionCallbackResultDropdown.tsx]: I can describe what it's used for but not what uses it. The latter is breaking encapsulation and bound to get out of date. Someone can easily import the file in another location and would not know to update the comment and it would be misleading.

Similar to my complaints about Lars' comments. Adding comments doesn't affect how you navigate code it only adds extra information to that code. It may better inform you but not change how the editors link and navigate files.  If you're using "find" for a certain comment name you think is next to a piece of code, instead use "find" for the actual code.
